File: vuln-sandbox/typescript/cwe-78/command-injection.ts
// VULNERABLE: InjectChildProcessShellTrue — spawn with shell:true (TypeScript)
// Rule: InjectChildProcessShellTrue | CWE-78 | Severity: CRITICAL

import express, { Request, Response } from 'express';
import { spawn } from 'child_process';

const app = express();
app.use(express.json());

interface RunRequest {
  command: string;
  args?: string[];
}

app.post('/run', (req: Request<{}, {}, RunRequest>, res: Response): void => {
  const { command, args = [] } = req.body;

  // VULNERABLE: shell:true allows shell metacharacter injection
  // An attacker can pass: "ls" with args ["; cat /etc/passwd"] to read sensitive files
  const proc = spawn(command, args, { shell: true });

  let output = '';
  proc.stdout.on('data', (data: Buffer) => { output += data.toString(); });
  proc.stderr.on('data', (data: Buffer) => { output += data.toString(); });
  proc.on('close', (code: number | null) => {
    res.json({ output, exitCode: code });
  });
});

app.listen(3000);

File: vuln-sandbox/java/cwe-78/CommandInjection.java
// VULNERABLE: java-cmdi-runtime-exec-concat — Runtime.exec with string concatenation
// Rule: java-cmdi-runtime-exec-concat | CWE-78 | Severity: CRITICAL

import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.BufferedReader;
import java.io.InputStreamReader;

public class CommandInjection extends HttpServlet {

    @Override
    protected void doGet(HttpServletRequest request, HttpServletResponse response) {
        String hostname = request.getParameter("host");

        try {
            // VULNERABLE: user-controlled input concatenated into shell command
            // An attacker can pass: localhost; cat /etc/passwd to read sensitive files
            Process proc = Runtime.getRuntime().exec("ping -c 1 " + hostname);

            BufferedReader reader = new BufferedReader(
                new InputStreamReader(proc.getInputStream())
            );

            String line;
            StringBuilder output = new StringBuilder();
            while ((line = reader.readLine()) != null) {
                output.append(line).append("\n");
            }

            response.getWriter().println(output.toString());
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}
